crates/net/src/http/response.rs

```diff
@@ -78,43 +93,93 @@ impl Response {
-    pub async fn form_data(&self) -> Result<web_sys::FormData, Error> {
+    /// Returns the underlying body as [`web_sys::FormData`].
+    ///
+    /// # Note
+    ///
+    /// This consumes the request, if you need to access the body multiple times,
+    /// you should `clone` the request first.
+    ///
+    /// # Errors
+    ///
+    /// Throws a "TypeError" if the content type of the request is not `"multipart/form-data"` or
+    /// `"application/x-www-form-urlencoded"`.
+    ///
+    /// Throws a "TypeError" if the body cannot be converted to [`web_sys::FormData`].
+    pub async fn form_data(self) -> Result<web_sys::FormData, Error> {
         let promise = self.0.form_data().map_err(js_to_error)?;
         let val = JsFuture::from(promise).await.map_err(js_to_error)?;
         Ok(web_sys::FormData::from(val))
     }
 
-    /// Reads the response to completion, parsing it as JSON.
-    #[cfg(feature = "json")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
-    pub async fn json<T: DeserializeOwned>(&self) -> Result<T, Error> {
-        serde_json::from_str::<T>(&self.text().await?).map_err(Error::from)
-    }
-
-    /// Reads the response as a String.
-    pub async fn text(&self) -> Result<String, Error> {
+    /// Returns the underlying body as a string.
+    ///
+    /// # Note
+    ///
+    /// This consumes the response, if you need to access the body multiple times,
+    /// you should `clone` the response first.
+    ///
+    /// # Errors
+    ///
+    /// This will return an error if the body cannot be decoded as utf-8.
+    pub async fn text(self) -> Result<String, Error> {
         let promise = self.0.text().unwrap();
         let val = JsFuture::from(promise).await.map_err(js_to_error)?;
         let string = js_sys::JsString::from(val);
         Ok(String::from(&string))
     }
 
+    /// Returns the underlying body as a string.
+    ///
+    /// # Note
+    ///
+    /// This consumes the response, if you need to access the body multiple times,
+    /// you should `clone` the response first.
+    ///
+    /// # Errors
+    ///
+    /// This will return an error if the body cannot be decoded as utf-8.
+    #[cfg(feature = "json")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+    pub async fn json<T: DeserializeOwned>(self) -> Result<T, Error> {
+        serde_json::from_str::<T>(&self.text().await?).map_err(Error::from)
+    }
+
     /// Gets the binary response
     ///
     /// This works by obtaining the response as an `ArrayBuffer`, creating a `Uint8Array` from it
     /// and then converting it to `Vec<u8>`
-    pub async fn binary(&self) -> Result<Vec<u8>, Error> {
+    ///
+    /// # Note
+    ///
+    /// This consumes the response, if you need to access the body multiple times,
+    /// you should `clone` the response first.
+    ///
+    /// # Errors
+    ///
+    /// This method may return a "RangeError"
+    pub async fn binary(self) -> Result<Vec<u8>, Error> {
```

Comment:
Methods that consume Request/Response bodies now take ownership of `self`. This makes `clone` infallible and prevents the class of bugs where users read the body multiple times.